fix: cert auth method watches cert file change and NewCreds() notific… (#28126)
hashicorp/vault: website/content/docs/agent-and-proxy/autoauth/methods/cert.mdx

- `client_key` `(string: optional)` - Path on the local disk to a single
  PEM-encoded private key matching the client certificate from client_cert.

- `reload` `(bool: optional, default: false)` - If true, causes the local x509
  key-pair to be reloaded from disk on each authentication attempt. This is useful
  in situations where client certificates are short-lived and automatically renewed.

>   Note that `enable_reauth_on_new_credentials` for autoauth needs to be enabled together.

Reviewer: Could you elaborate more on why `enable_reauth_on_new_credentials` is needed? I'm quite wary of exposing this if we don't need to. I'm also a little confused why this isn't needed in any of the tests that have been added (do we need additional coverage?) and this wasn't mentioned in the description of the PR.

If there's truly no other option for this use case to work than exposing such a config, I'd be open to it, but I'd want to remove the deprecation notice and also add a new changelog for this feature, along with why. First, though -- I'd like to better understand why we need this and what the behaviour would be if this was set to false.

Reviewer: It would also be important to document what the behaviour would be if this were set to false.

Author: Hi Violet
Thanks for the comments.
Coming to this "hidden" option, there are also other discussions/requirements from users like this: https://github.com/hashicorp/vault/issues/12233#issuecomment-2016340946

I know there is another way to signal proxy/agent to reload the entire config by using `SIGHUP`, but this is a bit troublesome if the cert provider is another service and it doesn't know who is using that on the node.

However, for `auth/cert`, the `reload` option is not a real reloading feature. Instead, it only guarantees when the main loop in auth handler has a chance for a next loop it reloads the fresh files from disk or just go with a cached HTTP client. And there's no way to do a real reloading without this hidden option which situation is similar to `auth/jwt`.

Certainly, we can leave this option to users who know this trick and keep it remain undocumented.
Wdyt?

Reviewer: Hmm, from what I can tell in the git blame, this has been there for six years, and we haven't wanted to remove it yet. I think I'm convinced. Let's document it. 

We should delete the two bits of text in Agent/Proxy config that say the following:
```
	// NOTE: This is unsupported outside of testing and may disappear at any
	// time.
```

I'll request a couple more docs updates around this area too, but I'm okay with going ahead with making the formerly undocumented feature real, especially as people seem to want it for other use cases. Thanks for the well-considered response!

Author: > this has been there for six years

That is really awesome ;)

Do you know the author of it? Not sure if there's any security concern when enabling it. But since it's disabled by default, things are still under control and users can opt-in to it.

Reviewer: There's no security concern -- users could still use it before, it was just added for testing purposes, it seems.

> - `reload` `(bool: optional, default: false)` - If true, causes the local x509
>   key-pair to be reloaded from disk on each authentication attempt. This is useful
>   in situations where client certificates are short-lived and automatically renewed.
>   Note that `enable_reauth_on_new_credentials` for autoauth needs to be enabled together.
>   See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).

Reviewer: ```suggestion
- `reload` `(bool: optional, default: false)` - If true, causes the local x509
  key-pair to be reloaded from disk on each authentication attempt. This is useful
  in situations where client certificates are short-lived and automatically renewed.
  Note that `enable_reauth_on_new_credentials` for auto-auth will need to be additionally enabled
  for immediate re-auth on a new certificate. See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).
```

> - `reload_period` `(duration: "1m", optional)` - The duration after which the Agent

Reviewer: ```suggestion
- `reload_period` `(duration: "1m", optional)` - The duration after which auto-auth
```

Reviewer: auto-auth is supported for both Agent and Proxy, so best to be generic

Author: Agree, changed it

- `client_key` `(string: optional)` - Path on the local disk to a single
  PEM-encoded private key matching the client certificate from client_cert.

- `reload` `(bool: optional, default: false)` - If true, causes the local x509
  key-pair to be reloaded from disk on each authentication attempt. This is useful
  in situations where client certificates are short-lived and automatically renewed.
  Note that `enable_reauth_on_new_credentials` for auto-auth will need to be additionally
  enabled for immediate re-auth on a new certificate.
  See [Auto-Auth Configuration](/vault/docs/agent-and-proxy/autoauth#configuration).

> - `reload_period` `(duration: "1m", optional)` - The duration after which the auto-auth

Reviewer: ```suggestion
- `reload_period` `(duration: "1m", optional)` - The duration after which auto-auth
```
